### ISSUE TITLE
Not working in Fedora 40

### ISSUE THREAD

**Issue by sauliaka** [morrownr/8821au-20210708#132]
Hi,

I happen to have a wifi dongle with rtl8811au chipset and I can't get this driver to work in it. I'm not that experienced what it comes to coding or scripting, but running ./install-driver.sh seems to give an error related to Linux kernel headers missing in latest Fedora for starters. This is probably because of the fresh (6.x+) kernel version. I was wondering if it's possible to fix it somehow, as the adapter works 100% correctly in Linux Mint with 5.15.x kernel.

Thanks in advance, if you have time to check it out.

**Comment by sunwire** [comment]
You have to install kernel-devel package

**Comment by sauliaka** [comment]
> You have to install kernel-devel package

is it already installed.

**Comment by mothsarecool524** [comment]
> > You have to install kernel-devel package
> 
> is it already installed.

By default, no. And installing it doesn't fix the issue.

**Comment by sauliaka** [comment]
> > > You have to install kernel-devel package
> > 
> > 
> > is it already installed.
> 
> By default, no. And installing it doesn't fix the issue.

I personally installed it before running that script but still there comes error. And you're correct, installing it didn't fix the issue :)

**Comment by morrownr** [comment]
Hi @sauliaka 

> This is probably because of the fresh (6.x+) kernel version. 

This driver is working on all 6.x kernels that have been released and it is working on 6.9 that has not been released yet.

> I was wondering if it's possible to fix it somehow, as the adapter works 100% correctly in Linux Mint with 5.15.x kernel.

It may be possible to fix this but this is problem in your system, not with this driver as far as I can tell. You can rerun `sudo sh install-driver.sh` and copy the screen output and post it here. If the problem is indeed missing header files, that is what needs to be fixed. The best place to ask about this is probably a fedora forum.

**Comment by mothsarecool524** [comment]
> Hi @sauliaka 
> 
> > This is probably because of the fresh (6.x+) kernel version. 
> 
> This driver is working on all 6.x kernels that have been released and it is working on 6.9 that has not been released yet.
> 
> > I was wondering if it's possible to fix it somehow, as the adapter works 100% correctly in Linux Mint with 5.15.x kernel.
> 
> It may be possible to fix this but this is problem in your system, not with this driver as far as I can tell. You can rerun `sudo sh install-driver.sh` and copy the screen output and post it here. If the problem is indeed missing header files, that is what needs to be fixed. The best place to ask about this is probably a fedora forum.
> 
> 

![Screenshot_20240428_154622.png](https://github.com/morrownr/8821au-20210708/assets/124483129/7f54b626-e069-4463-8cd1-24ac42bb3519)

I'm having the same issue here, I installed `kernel-devel` to see if that fixed it, but it did not, in my case. Same version of Fedora, 40, on the KDE Plasma spin. Relatively fresh install as well.

**Comment by sauliaka** [comment]
> > Hi @sauliaka
> > > This is probably because of the fresh (6.x+) kernel version.
> > 
> > 
> > This driver is working on all 6.x kernels that have been released and it is working on 6.9 that has not been released yet.
> > > I was wondering if it's possible to fix it somehow, as the adapter works 100% correctly in Linux Mint with 5.15.x kernel.
> > 
> > 
> > It may be possible to fix this but this is problem in your system, not with this driver as far as I can tell. You can rerun `sudo sh install-driver.sh` and copy the screen output and post it here. If the problem is indeed missing header files, that is what needs to be fixed. The best place to ask about this is probably a fedora forum.
> 
> ![Screenshot_20240428_154622.png](https://private-user-images.githubusercontent.com/124483129/326302106-7f54b626-e069-4463-8cd1-24ac42bb3519.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTQzNDM3MTUsIm5iZiI6MTcxNDM0MzQxNSwicGF0aCI6Ii8xMjQ0ODMxMjkvMzI2MzAyMTA2LTdmNTRiNjI2LWUwNjktNDQ2My04Y2QxLTI0YWM0MmJiMzUxOS5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwNDI4JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDQyOFQyMjMwMTVaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT04ZDgxNDA5ZjVhMWRjOTU5NWMxYzUxNTYzOTRkNWQwYTg1ZGY0NzI3ZTc1MmMzZTM4Y2NiMjc3YzgxMzI0YzhjJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.oI2WLXcRXXaiCSgzamJTy1hYlVajx71X8bU5DYOpeW0)
> 
> I'm having the same issue here, I installed `kernel-devel` to see if that fixed it, but it did not, in my case. Same version of Fedora, 40, on the KDE Plasma spin. Relatively fresh install as well.

This is exactly the same problem I had.

**Comment by morrownr** [comment]
You only posted the error. What I want you to so is post the entire screen output of `sudo sh install-driver.sh` and please post it as text.

**Comment by mothsarecool524** [comment]
`Your kernel header files aren't properly installed.
Please consult your distro documentation or user support forums.
Once the header files are properly installed, please run "sudo ./install-driver.sh"`

This is the output of the aforementioned command.

**Comment by morrownr** [comment]
Give me the output of:

$ uname -r

**Comment by mothsarecool524** [comment]
`6.8.7-300.fc40.x86_64`

**Comment by morrownr** [comment]
Run the following:

$ sudo sh remove-driver.sh
$ make clean
$ make
$ sudo make install
$ sudo reboot

**Comment by mothsarecool524** [comment]
Seems to have worked successfully, now.

**Comment by morrownr** [comment]
You are welcome but now I need your help if you have time.

If I deactivate the code to detect if the header files are installed, can I get you to test it?

I need to find out what is going on in fedora that my code is not working properly. Using `sudo sh install-driver.sh` is much better than the manual installation that I showed you because it will automatically recompile the driver for you when a new kernel flows in.

**Comment by mothsarecool524** [comment]
> You are welcome but now I need your help if you have time.
> 
> If I deactivate the code to detect if the header files are installed, can I get you to test it?
> 
> I need to find out what is going on in fedora that my code is not working properly. Using `sudo sh install-driver.sh` is much better than the manual installation that I showed you because it will automatically recompile the driver for you when a new kernel flows in.

Sure! :D

Though another issue just popped up for me. The networks are listed with each wifi adapter name next to them, but the network I want to connect do isn't connecting on the USB adapter. (Incorrect password, even though password is correct. Network is configured correctly too.)

**Comment by morrownr** [comment]
Okay. Run the following commands in the driver directory:

$ git pull
$ make clean
$ sudo sh install-driver,sh

Make sure that you reboot. See if it works.

Then we can look for the connection problem.

**Comment by mothsarecool524** [comment]
It now works just fine with no issues this time. Don't know what could've been causing them, hopefully these issues don't appear for Fedora 40 users that haven't installed this driver before.

**Comment by mothsarecool524** [comment]
@sauliaka Have you tried these fixes? Should work now, just `git pull` in the `~/src/` directory that was made during initial installation, and then go from there with the latest instructions from morrownr's latest comment.

**Comment by sauliaka** [comment]
I managed to install it now too, wohoo :)

**Comment by morrownr** [comment]
> It now works just fine with no issues this time.

> I managed to install it now too, wohoo :)

Good for both of you but not so good for me, Now I need to find out where fedora is installing its headers files. I have a lot of routines in the `install-driver.sh` script that are designed to find and fix problems or in this case, let you know there is a problem that you need to fix. Inactivating this code is solved your problem but the code was there for a good reason... header files not being installed is a big problem. Obviously your header files are installed but they are not in the standard location, which is what I check for. 

Normally header files are installed in the following location:

/lib/modules/$(uname -r)/build

Where are your header files?

**Comment by sauliaka** [comment]
> Normally header files are installed in the following location:
> 
> /lib/modules/$(uname -r)/build
> 
> Where are your header files?

I believe there were 2 different versions under /lib/modules, 6.8.5-301.fc40 and 6.8.7-300.fc40. Under 6.8.5-301.fc40 build link was broken when pointing to under /usr/src/kernels/6.8.5-301.fc40, that directory didn't exist and your installer script tried to use that version. I'm here currently kinda swapping ssd drives as I'm doing some testing with my older laptop, but I can try to confirm this for you later.

**Comment by laurent-onenano** [comment]
Hi everyone,

I've followed all instructions, installed, uninstalled, checked other drivers, ... Nothing worked so far.

I'm currently under Fedora 40 with kernel 6.8.11-300.fc40.x86_64
I've unfortunately bought a TP-LINK AC600 Archer T2U Nano (USB)

make (1/2) (please note that the difference in compiler in only after the last upgrade)
![image](https://github.com/morrownr/8821au-20210708/assets/124622149/2475cdd6-e809-4061-a0b7-58082cca7f1c)

make (2/2)
![image](https://github.com/morrownr/8821au-20210708/assets/124622149/16bc1296-316a-437a-8ce4-7eabfc3e322e)

make install
![image](https://github.com/morrownr/8821au-20210708/assets/124622149/6bdc1c02-b2c0-4588-9e1d-be8d0d91ea99)

Module is located at /lib/modules/6.8.11-300.fc40.x86_64/kernel/drivers/net/wireless/8821au.ko
![image](https://github.com/morrownr/8821au-20210708/assets/124622149/f15bbd52-681a-403e-8374-35b9e26a3b0b)

inxi -aN
![image](https://github.com/morrownr/8821au-20210708/assets/124622149/67584608-837b-4157-9571-abbc2a980fda)

lsusb
![image](https://github.com/morrownr/8821au-20210708/assets/124622149/e7586e2d-b3db-4acd-9928-487cfb53839d)

lsusb -t (you can see there is no driver enabled)
![image](https://github.com/morrownr/8821au-20210708/assets/124622149/41a1a4a9-337f-4974-bf36-0d1cb5684bbf)

And still, no Wi-Fi adapter found.
![image](https://github.com/morrownr/8821au-20210708/assets/124622149/7584cc3e-f5dc-40e5-b5ec-40cdf088573e)


Does anyone have any idea?

**Comment by sunwire** [comment]
> Does anyone have any idea?

Is module loaded? (lsmod|grep  8821au)
if not try to load manually

**Comment by laurent-onenano** [comment]
Good point!

It was not loaded because of SecureBoot I guess

![image](https://github.com/morrownr/8821au-20210708/assets/124622149/3505b45c-8bdd-4bce-9943-87e559acbd00)

![image](https://github.com/morrownr/8821au-20210708/assets/124622149/9151ac91-bff5-4bbe-873a-3474bd31d088)

I've tried to use the signed install but I do not have "sign" and package "sign-file" is non-existent in my repos. However, bash suggested "obs-signd" which I downloaded.

![image](https://github.com/morrownr/8821au-20210708/assets/124622149/0eca62db-61e3-4bed-b1d6-5688b15645d7)

make sign-install didn't work.
![image](https://github.com/morrownr/8821au-20210708/assets/124622149/4e5ed834-e5df-42a0-8d53-7da387f8d18a)

I've done the 2 steps independently.
![image](https://github.com/morrownr/8821au-20210708/assets/124622149/f728b80a-78a6-47e0-ba28-150e1d345f27)

![image](https://github.com/morrownr/8821au-20210708/assets/124622149/08b8b78d-73f6-4454-abb0-9e7edcd0c72a)

I'm rebooting now and will check if module is loaded

**Comment by laurent-onenano** [comment]
Same problem

![image](https://github.com/morrownr/8821au-20210708/assets/124622149/418ee547-7907-413e-9244-dabf0c5fad9c)

**Comment by morrownr** [comment]
@laurent-onenano 

An alternative to debugging the problems with the driver in this repo is to help us test a new driver that is intended to be upstreamed into the Linux kernel soon. You can get information in the following issue:

https://github.com/morrownr/8821au-20210708/issues/133

Read the first message carefully, uninstall this driver and then follow the instructions in the README of the rtw88 repo. We know the driver is in good shape with only a few minor issues left. Your testing would be most welcome and you may find that it works better than the the driver in this repo.

Let me know how it goes.

@morrownr

**Comment by laurent-onenano** [comment]
Happy to help as well mate

**Comment by morrownr** [comment]
@laurent-onenano 

Great. The invitation goes to @sunwire and the others in this thread as well.

If you have any questions about installing the new rtw88 driver, let me know. I know it is a rather manual process for now but your adapters will be Plug and Play once we get it into the kernel.

**Comment by sunwire** [comment]
> Same problem

If secure boot is enabled you have to:

1. Sign the module
[example](https://access.redhat.com/documentation/en-us/red_hat_enterprise_linux/9/html/managing_monitoring_and_updating_the_kernel/signing-a-kernel-and-modules-for-secure-boot_managing-monitoring-and-updating-the-kernel#signing-kernel-modules-with-the-private-key_signing-a-kernel-and-modules-for-secure-boot)
2. Enroll the key

To enroll the key use command
sudo mokutil --import /path/to/public_key

4. Reboot

After restarting, a menu will appear in which you will need to enroll the key

It may helps https://youtu.be/rlCSVyv54fc
PS1. I suggest using DKMS (It will automatically build and sign the module)
PS2. If you wish you can use my COPR repo https://copr.fedorainfracloud.org/coprs/sunwire/dkms-8821au/

**Comment by laurent-onenano** [comment]
It works fine!

I had another problem and it looks like the nvidia driver (from nvidia repo) created a conflict with MOK. The screen didn't appear anymore. I uninstalled the driver and all worked fine.

Thanks guys!

**Comment by morrownr** [comment]
@laurent-onenano 

That's good to hear. Leave this issue open in case someone else runs into this.

You are still welcome to test the new rtw88 driver I mentioned above. The more testing we get, the sooner we can upstream it and then you won't have to compile anything.
